### PR TITLE
[#53] Setup unit testing

### PR DESCRIPTION
**Pull request by luongvo**
https://github.com/luongvo/flutter-survey/issues/53

## What happened 👀

Setup all the required components for unit testing.
 
## Insight 📝

- [x] The environment should be ready for writing tests along with the feature development progress.
- [x] Add unit tests for existing components:
  - [x] OauthRepository
  - [x] LoginUseCase
  - [x] LoginViewModel
- [x] Integrate [codecov](https://about.codecov.io/) for code coverage report.
 
## Proof Of Work 📹

See codecov report below 👇

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,17 +11,26 @@ A new Flutter project.
...
+- Create these env files in the root directory according to the flavors and add the required
+  environment variables into them. The example environment variable is in `.env.sample`.
 
   - Staging: `.env.staging`
 
   - Production: `.env`
 
-- Generate project assets: `$ flutter packages pub run build_runner build --delete-conflicting-outputs`
+- Generate project
+  assets: `$ flutter packages pub run build_runner build --delete-conflicting-outputs`
```
Comment (manh-t) [inline README.md:22]: @luongvo oh, are you intending to make a line break here?

File: lib/api/request/oauth_token_request.dart [hunk 2]
```diff
@@ -7,15 +7,15 @@ const String GRANT_TYPE_PASSWORD = "password";
 @JsonSerializable()
 class OAuthTokenRequest {
   @JsonKey(name: 'grant_type')
-  String grantType;
+  final String grantType;
```
Comment (manh-t) [inline lib/api/request/oauth_token_request.dart:10]: making a variable as `final` is a good habit 👍

File: lib/local/shared_preference_helper.dart [hunk 3]
```diff
@@ -1,10 +1,10 @@
 import 'package:injectable/injectable.dart';
 import 'package:shared_preferences/shared_preferences.dart';
 
-const String _PREF_KEY_TYPE = 'PREF_KEY_TYPE';
-const String _PREF_KEY_ACCESS_TOKEN = 'PREF_KEY_TOKEN';
-const String _PREF_KEY_REFRESH_TOKEN = 'PREF_KEY_REFRESH_TOKEN';
-const String _PREF_KEY_TOKEN_EXPIRATION = 'PREF_KEY_TOKEN_EXPIRATION';
+const String PREF_KEY_TYPE = 'PREF_KEY_TYPE';
```
Comment (manh-t) [inline lib/local/shared_preference_helper.dart:4]: 👍

**Review by manh-t — COMMENTED** [review]
The test cases look good 💯

File: test/usecase/login_use_case_test.dart [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+      expect(result, isA<Success>());
+    });
+
+    test(
+        'When logging in with incorrect email or password, it returns Failed result',
+        () async {
+      when(mockRepository.login(
+              email: anyNamed('email'), password: anyNamed('password')))
+          .thenAnswer((_) => Future.error(
```
Comment (sleepylee) [inline test/usecase/login_use_case_test.dart:45]: I think you can use thenThrow to avoid creating a Future.error

**Review by sleepylee — COMMENTED** [review]
similar suggestion to the other places 😃

**Comment by manh-t** [comment]
Only the comment from Trung, rest seems to be good to me 👍

**Review by sleepylee — APPROVED** [review]

File: test/api/repository/oauth_repository_test.dart [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          requestOptions: RequestOptions(path: '')));
+      try {
+        await oauthRepository.login(
+          email: "email",
+          password: "password",
+        );
+      } catch (e) {
+        expect(e, isInstanceOf<NetworkExceptions>());
+      }
+    });
```
Comment (sleepylee) [inline test/api/repository/oauth_repository_test.dart:77]: I think we can agree to write in this way which is cleaner than putting try/catch in a test 🤔 

```
final result = () => oauthRepository.login(
        email: "email",
        password: "password",
      );

      expect(result, throwsA(isA<NetworkExceptions>()));
```

Comment (luongvo) [reply test/api/repository/oauth_repository_test.dart:77]: oh, this should be shorter and cleaner for sure 👍

File: test/usecase/login_use_case_test.dart (hunk 4 above)
Comment (sleepylee) [reply test/usecase/login_use_case_test.dart:45]: I think moving forward we better decide to use try/catch or Future/then/error to be consistent. I found that I was wrong about using `func.then().error()` before too 😓  I'm checking if this is why the catch error by using `error()` block wasn't called

https://dart.dev/guides/libraries/futures-error-handling

image

Comment (sleepylee) [reply test/usecase/login_use_case_test.dart:45]: they are the same imo 🤔  Future = async func {} . Await = then (). catch = catchError 🤷

Comment (sleepylee) [reply test/usecase/login_use_case_test.dart:45]: similar issue here: https://stackoverflow.com/questions/63493407/dart-future-catcherror-not-called-in-unit-test 😔

Comment (luongvo) [reply test/usecase/login_use_case_test.dart:45]: omg, I thought you got why I can't apply `thenThrow` here already 🥲 We need to mock with Future here, the one in Repository is different.

**Review by sleepylee — APPROVED** [review]

**Pull request merged**